File: 3445.maximum-difference-between-even-and-odd-frequency-ii.cpp
#
# @lc app=leetcode id=3445 lang=cpp
#
# [3445] Maximum Difference Between Even and Odd Frequency II
#
# @lc code=start
class Solution {
public:
    int maxDifference(string s, int k) {
        int n = s.size();
        unordered_map<char, int> freq;
        int maxDiff = INT_MIN;
        int start = 0;
        
        for (int end = 0; end < n; ++end) {
            freq[s[end]]++;
            
            // When we have a valid window of at least size k
            if (end - start + 1 >= k) {
                // Check for odd/even frequency characters within this window
                int oddMax = INT_MIN, evenMax = INT_MIN;
                for (auto &p : freq) {
                    if (p.second % 2 == 1) { // odd frequency
                        oddMax = max(oddMax, p.second);
                    } else if (p.second > 0) { // non-zero even frequency
                        evenMax = max(evenMax, p.second);
                    }
                }
                
                // Compute potential max difference
                if (oddMax != INT_MIN && evenMax != INT_MIN) {
                    maxDiff = max(maxDiff, oddMax - evenMax);
                }
                
                // Slide the window by reducing start point
                freq[s[start]]--;
                if (freq[s[start]] == 0) {
                    freq.erase(s[start]);
                }
                start++;
            }
        }
        
        return maxDiff == INT_MIN ? -1 : maxDiff;
    }
};
# @lc code=end

File: 3445.maximum-difference-between-even-and-odd-frequency-ii.java
#
# @lc app=leetcode id=3445 lang=java
#
# [3445] Maximum Difference Between Even and Odd Frequency II
#

# @lc code=start
class Solution {
    public int maxDifference(String s, int k) {
        int n = s.length();
        int maxDiff = Integer.MIN_VALUE;
        boolean found = false;
        // Enumerate all digit pairs (a, b), a != b
        for (int a = 0; a < 5; a++) {
            for (int b = 0; b < 5; b++) {
                if (a == b) continue;
                int[] freq = new int[5];
                int left = 0;
                for (int right = 0; right < n; right++) {
                    freq[s.charAt(right) - '0']++;
                    // Slide window to maintain length at least k
                    while (right - left + 1 > k) {
                        freq[s.charAt(left) - '0']--;
                        left++;
                    }
                    if (right - left + 1 >= k) {
                        int fa = freq[a];
                        int fb = freq[b];
                        if (fa % 2 == 1 && fb > 0 && fb % 2 == 0) {
                            maxDiff = Math.max(maxDiff, fa - fb);
                            found = true;
                        }
                    }
                }
            }
        }
        return found ? maxDiff : -1;
    }
}
# @lc code=end